hack/self-hosted-registry-demo/demo.sh

```diff
@@ -0,0 +1,270 @@
+readonly SCRIPT_DIR
+
+GIT_ROOT="$(git rev-parse --show-toplevel)"
+readonly GIT_ROOT
+
+MINDTHEGAP_VERSION="$(gojq -r .version "${GIT_ROOT}/dist/metadata.json")"
+readonly MINDTHEGAP_VERSION
+
+IMAGE_VERSION="${MINDTHEGAP_VERSION}-$(go env GOARCH)"
+readonly IMAGE_VERSION
+
+# Build the image bundle
+"${GIT_ROOT}/dist/mindthegap_$(go env GOOS)_$(go env GOARCH)/mindthegap" \
+  create bundle \
+  --images-file "${SCRIPT_DIR}/bootstrap-images.txt" \
+  --output-file "${SCRIPT_DIR}/bundle.tar" \
+  --platform linux/"$(go env GOARCH)" \
+  --image-pull-concurrency 20 \
+  --overwrite
+
+export KUBECONFIG="${SCRIPT_DIR}/kubeconfig"
+
+# Create the kind cluster, mounting the bundle into the control-plane node of the KinD cluster
+kind create cluster --name self-hosted-registry-demo --config <(
+  cat <<EOF
+kind: Cluster
+apiVersion: kind.x-k8s.io/v1alpha4
+nodes:
+  - role: control-plane
+    extraMounts:
+      - hostPath: "${SCRIPT_DIR}/bundle.tar"
+        containerPath: /registry-data/bundle.tar
+  - role: worker
+containerdConfigPatches:
+  - |-
+    [plugins."io.containerd.grpc.v1.cri".registry]
+      config_path = "/etc/containerd/certs.d"
+networking:
+  # the default CNI will not be installed
+  disableDefaultCNI: true
+EOF
+)
+
+# Show that the nodes are not Ready - no CNI has been installed yet (default kindnet has been disabled above).
+kubectl get nodes
+
+# Load the necessary images into the KinD cluster - this could be done by building the KinD node image
+# with the necessary images, but this is a simpler way to demonstrate the concept.
+kind load docker-image --name self-hosted-registry-demo --quiet \
+  ko.local/{mindthegap,wait-for-file-to-exist,copy-file-to-pod}:"${IMAGE_VERSION}"
+
+# Configure containerd on each node to use the in-cluster registry service address - this will
+# ensure that the images are only pulled from the in-cluster registry service.
+readonly DEFAULT_REGISTRY_DIR="/etc/containerd/certs.d/_default"
+for node in $(kind get nodes --name self-hosted-registry-demo); do
+  docker container exec "${node}" mkdir -p "${DEFAULT_REGISTRY_DIR}"
+  cat <<EOF | docker container exec -i "${node}" tee "${DEFAULT_REGISTRY_DIR}/hosts.toml" >/dev/null
+server = "http://10.96.0.20"
+
+# See https://github.com/containerd/containerd/issues/10027 for containerd bug that requires this workaround,
+# having an empty host tree here.
+[host]
+EOF
+done
+
+# Create the registry pod.
+cat <<EOF | kubectl apply --server-side -n kube-system -f -
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: kube-system
+  name: temporary-registry
+  labels:
+    app: in-cluster-registry
+spec:
+  tolerations:
+    - operator: Exists
+  affinity:
+    nodeAffinity:
+      requiredDuringSchedulingIgnoredDuringExecution:
+        nodeSelectorTerms:
+          - matchExpressions:
+              - key: node-role.kubernetes.io/control-plane
+                operator: Exists
+  hostNetwork: true
+  initContainers:
```

Comment:
Im curious why you went with this approach of copying files from a job instead of using `ko.local/copy-file-to-pod:${IMAGE_VERSION}` here directly and copy the bundle to a shared volume?

Reply:
The bundle we're copying is large, maybe about 15GB. I do not want to bake that into the machine images used for the nodes. We have to copy it from the bootstrap machine where the use will have it present (but also not in the bootstrap image, but just locally and mounted into the bootstrap container).